librelane: init at 3.0.0.dev47-unstable-2026-01-16

gonsolo opened the pull request:
ASIC implementation flow infrastructure, successor to OpenLane
https://github.com/librelane/librelane

- Built on platform:
  - [x] x86_64-linux
- [x] Tested basic functionality of all binary files, usually in `./result/bin/`.

Dvd-Znf reviewed (commented):
i think you accidentally included commits other than `librelane: init at 3.0.0.dev45` by maybe branching from anothers PR branch instead of clean master

you can try using git rebase -i to remove the other commits

gonsolo commented:
> i think you accidentally included commits other than `librelane: init at 3.0.0.dev45` by maybe branching from anothers PR branch instead of clean master

All the commits are dependencies of librelane. I rebased onto them to be able to nix build librelane before all of them are upstream.

I converted this PR to a draft for now.

gonsolo commented:
>  (`librelane --smoke-test` must pass.)

It does now.

> The overlays we use to get LibreLane working entirely in Nix are not up to nixpkgs standards and they would take a monumental effort to get them there. That's why I haven't bothered myself. https://github.com/librelane/librelane/blob/c68e32b888bb5f42fec75fa0d8bcc7fb8c994ab6/default.nix

I didn't look very close at those.

@donn Can you take a look whether I forgot something? As mentioned above, the smoke test works for me now.

donn commented:
If `--smoke-test` passes I have no objections. Though I do wonder how it passes considering the OpenROAD and Yosys Python interpreters shouldn't have access to the `click` library according to what I'm seeing here…? Attempting to import either library typically causes an immediate crash with this setup.

donn commented on pkgs/by-name/li/librelane/package.nix:
```diff
@@ -0,0 +1,101 @@
+
+  postFixup = ''
+    wrapProgram $out/bin/librelane \
+      --set PYTHONPATH "$PYTHONPATH" \
```
Oh, this would fix the yosys/openroad interpreter issue. The difference being trying to use `python3 -m librelane --smoke-test` would still fail, as would attempting to use librelane as a library:

```
from librelane.flows import Flow

Classic = Flow.factory.get("Classic")

flow = Classic(
    {
        "PDK": "sky130A",
        "DESIGN_NAME": "spm",
        "VERILOG_FILES": ["./src/spm.v"],
        "CLOCK_PORT": "clk",
        "CLOCK_PERIOD": 10,
    },
    design_dir=".",
)

flow.start()
```

But this works considering LibreLane isn't in the Python module packageset.

---

The way I handle it upstream is by essentially replicating "withPackages" for the Yosys and OpenROAD binaries to ensure they get their own site-packages with version-appropriate click and pyyaml.

For Yosys, this could be very easily handled by switching to Pyosys if https://github.com/NixOS/nixpkgs/issues/23676 weren't somehow still a thing.

gonsolo replied:
> Completely overriding PYTHONPATH prevents users from being able to load plugins: https://librelane.readthedocs.io/en/latest/usage/writing_plugins.html

Done.

gonsolo commented:
> If `--smoke-test` passes I have no objections.

Cool, thanks.

> Though I do wonder how it passes considering the OpenROAD and Yosys Python interpreters shouldn't have access to the `click` library according to what I'm seeing here…? Attempting to import either library typically causes an immediate crash with this setup.

I see click in Yosys's package.nix, and grepping through OpenROAD seems to wanting to install it, but I can find no use.

donn commented:
Mostly LGTM, need to sit down and test it now. Really busy so might take a while.

gonsolo commented:
Dependencies (roughly):

[librelane](https://github.com/NixOS/nixpkgs/pull/471712):	openroad, netgen-vlsi,  yosys		(Draft)
[openroad](https://github.com/NixOS/nixpkgs/pull/477789):	lemon-graph, yosys			(Draft)
[yosys](https://github.com/NixOS/nixpkgs/pull/471062):		lemon-graph, openroad
[coloquinte](https://github.com/NixOS/nixpkgs/pull/477801):	lemon-graph
[netgen-vlsi](https://github.com/NixOS/nixpkgs/pull/477797):	-					(Draft)
[lemon-graph](https://github.com/NixOS/nixpkgs/pull/477756):	-

FYI @donn, @mole99, @eljamm.

gonsolo committed:
librelane: init at 3.0.0.dev47-unstable-2026-01-16

Co-authored-by: Mohamed Gaber <me@donn.website>
Co-authored-by: Gaétan Lepage <gaetan@glepage.com>

GaetanLepage commented on pkgs/by-name/li/librelane/package.nix:
```diff
@@ -0,0 +1,111 @@
+
+  nativeBuildInputs = [
+    python3Packages.poetry-core
+    python3Packages.pythonRelaxDepsHook
```
```suggestion
```
Not needed.

GaetanLepage commented on pkgs/by-name/li/librelane/package.nix:
```diff
@@ -0,0 +1,111 @@
+  ];
+
+  nativeBuildInputs = [
+    python3Packages.poetry-core
```
Should go into `build-system`.

gonsolo replied:
Done.

GaetanLepage commented on pkgs/by-name/li/librelane/package.nix:
```diff
@@ -0,0 +1,111 @@
+let
+  inherit (python3Packages) libparse-python;
+in
+python3Packages.buildPythonApplication {
```
```suggestion
python3Packages.buildPythonApplication (finalAttrs: {
```

gonsolo commented on pkgs/by-name/li/librelane/package.nix:
```diff
@@ -0,0 +1,111 @@
+  src = fetchFromGitHub {
+    owner = "librelane";
+    repo = "librelane";
+    rev = "f47ad8e8ef8d2abd648c79e4b7e32568f20951ea";
```
I used `rev = finalAttrs.rev;` instead of tag since we are using the latest commit from the dev branch, not the tagged version.

gonsolo commented on pkgs/by-name/li/librelane/package.nix:
```diff
@@ -0,0 +1,111 @@
+    hash = "sha256-TXM0oVOlfRZvNP6XObztpx30ddTiO4ije5WQS0+1bkM=";
+  };
+
+  propagatedBuildInputs = with python3Packages; [
```
Done.

gonsolo commented:
I think I addressed all your requests, @GaetanLepage. Thanks for the thorough review!

Smoke test is still passing, @donn @mole99:

```
* Antenna
Passed ✅

* LVS
Passed ✅

* DRC
Passed ✅

[12:48:54] INFO     Flow complete.                                                                                       sequential.py:411

...

[12:48:54] INFO     Smoke test passed.
```

SuperSandro2000 commented on pkgs/by-name/li/librelane/package.nix:
```diff
@@ -0,0 +1,111 @@
+    wrapProgram $out/bin/librelane \
+      --suffix PYTHONPATH : "$PYTHONPATH" \
+      --prefix PATH : ${
+        lib.makeBinPath [
```
If we only use those packages here, we do not need to add them to buildInputs and they would not do anything there.

SuperSandro2000 commented on pkgs/by-name/li/librelane/package.nix:
```diff
@@ -0,0 +1,111 @@
+      }
+  '';
+
+  catchConflicts = false;
```
Why not?

gonsolo commented on pkgs/by-name/li/librelane/package.nix:
```diff
@@ -0,0 +1,111 @@
+python3Packages.buildPythonApplication (finalAttrs: {
+  pname = "librelane";
+  version = "3.0.0.dev47-unstable-2026-01-12";
+  rev = "b38f7d33a4fb9f5fb8b99715e9d6595c35df0d98";
```
I guess you're suggesting moving rev into fetchFromGitHub. Done.

gonsolo commented on pkgs/by-name/li/librelane/package.nix:
```diff
@@ -0,0 +1,99 @@
+    (fetchpatch {
+      name = "fix-click-comatibility";
+      url = "https://github.com/librelane/librelane/pull/844/changes/5b63efc87b42dfaa131544f1f3d5f220118891a2.patch";
+      hash = "...";
```
I committed via iPhone. I will be on the desktop in the the afternoon again.

gonsolo commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 471712`
Commit: `2c326cb69a706e97de0c9954bf1e2a1199fa5475`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 2 packages built:</summary>
  <ul>
    <li>librelane</li>
    <li>librelane.dist</li>
  </ul>
</details>

GaetanLepage commented on pkgs/by-name/li/librelane/package.nix:
```diff
@@ -0,0 +1,99 @@
+          netgen-vlsi
+          verilator
+          ruby
+          klayout
```
Nit: sort.

GaetanLepage replied:
Nit: sort.

gonsolo commented:
> Maybe you already answered this, but wouldn't it work to fetch the `3.0.0.dev47` tag instead of this commit?

I would have to apply a second PR since that release:
https://github.com/librelane/librelane/pull/843
Whatever you prefer.

I expect the newer PR 844 to be merged before a dev48, after which I would immediately update librelane in Nix, dropping the PR or PRs.

It can be done before or after merging, I'm just committed to get it merged. :)

GaetanLepage commented:
The commit message and PR name should match the version:
`librelane: init at 3.0.0.dev47-unstable-2026-01-12`

gonsolo commented:
> The commit message and PR name should match the version: `librelane: init at 3.0.0.dev47-unstable-2026-01-12`

Done.

gonsolo commented:
> > The commit message and PR name should match the version: `librelane: init at 3.0.0.dev47-unstable-2026-01-12`
> 
> Done.

I changed to librelane: init at 3.0.0.dev47-unstable-2026-01-16, since the patch from today was added.

GaetanLepage reviewed (approved):
LGTM
